Userland/Libraries/LibAudio/MP3Loader.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2021, Arne Elster <arne@elster.li>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "MP3Loader.h"
+#include "MP3HuffmanTables.h"
+#include "MP3Tables.h"
+#include <LibCore/File.h>
+#include <LibCore/FileStream.h>
+
+namespace Audio {
+
+MP3LoaderPlugin::MP3LoaderPlugin(StringView path)
+    : m_file(Core::File::construct(path))
+{
+    if (!m_file->open(Core::OpenMode::ReadOnly)) {
+        m_error = LoaderError { String::formatted("Can't open file: {}", m_file->error_string()) };
+        return;
+    }
+
+    off_t file_size = 0;
+    if (!m_file->seek(0, Core::SeekMode::FromEndPosition, &file_size)) {
+        m_error = LoaderError { String::formatted("Could not seek in file.") };
+        return;
+    }
+    m_file_size = file_size;
+    if (!m_file->seek(0, Core::SeekMode::SetPosition)) {
+        m_error = LoaderError { String::formatted("Could not seek in file.") };
+        return;
+    }
+
+    m_input_stream = make<Core::InputFileStream>(*m_file);
+    if (!m_input_stream || m_input_stream->has_any_error()) {
+        m_error = LoaderError { String::formatted("Could not create input stream on file.") };
+        return;
+    }
+
+    m_bitstream = make<InputBitStream>(*m_input_stream);
+    if (!m_bitstream || m_bitstream->has_any_error()) {
+        m_error = LoaderError { String::formatted("Could not create bit stream on top of input stream") };
+        return;
+    }
+}
+
+MP3LoaderPlugin::~MP3LoaderPlugin()
+{
+    if (m_bitstream)
+        m_bitstream->handle_any_error();
+
+    if (m_input_stream)
+        m_input_stream->handle_any_error();
+}
+
+MaybeLoaderError MP3LoaderPlugin::initialize()
```

Comment:
I realize you are just following the pattern of FLAC... but if you are writing a new plugin.. I think we can improve upon the previous design. 

We try to avoid having objects in uninitialized / undefined states. So `MyClass::initialize()` functions are usually not desirable. Can we create this via a factory function instead, where the factory does the object validation before creation, and passes in any dependencies that could error out before the `MP3LoaderPlugin` is initialized?

Reply:
That would definitely be an improvement. I'm currently waiting for @kleinesfilmroellchen 's FlacLoader stream API PR so that I can add memory stream support for the fuzzer to work. After that I may try the architectural change if filmroellchen isn't faster :^)